velox/exec/prefixsort/PrefixSortEncoder.h

```diff
@@ -28,52 +28,166 @@ namespace facebook::velox::exec::prefixsort {
+/// starting at the address of it, would be '0xdd 0xcc 0xbb 0xaa'. If we store
+/// them into a buff, and reverse the bytes of the buff : [0xaa, 0xbb, 0xcc,
+/// 0xdd], and then we can compare two buffs from the first byte to last
+/// byte, the compare result is equal to value-compare.
+/// For any two unsigned integers, a < b <==> ~a > ~b so we invert bits when
+/// descending order.
 template <>
-FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encode(int64_t value, char* row) {
-  const auto v = __builtin_bswap64(static_cast<uint64_t>(value));
-  simd::memcpy(row, &v, sizeof(int64_t));
-  row[0] = flipSignBit(row[0]);
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    uint32_t value,
+    char* dest) const {
+  auto& v = *reinterpret_cast<uint32_t*>(dest);
+  v = __builtin_bswap32(value);
+  if (!ascending_) {
+    v = ~v;
+  }
 }
 
+/// Compare two positive signed integers: storage layout is as same as unsigned
+/// integer, their sign-bit are same, flip sign-bit do not change result.
+/// Compare two negative signed integers: -n = ~n + 1, we can trait ~n + 1 as an
+/// unsigned integer, so the logic is as same as unsigned integer, also flip
+/// sign-bit do not change result.
+/// Compare positive vs negative: flip sign-bit to promise that positive
+/// always bigger than negative.
 template <>
-FOLLY_ALWAYS_INLINE void PrefixSortEncoder::decode<int64_t>(char* row) {
-  row[0] = flipSignBit(row[0]);
-  const auto v = __builtin_bswap64(*reinterpret_cast<uint64_t*>(row));
-  simd::memcpy(row, &v, sizeof(int64_t));
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    int32_t value,
+    char* dest) const {
+  encodeNoNulls((uint32_t)(value ^ 1u << 31), dest);
 }
 
-/// For testing only.
-namespace {
-template <typename T>
-FOLLY_ALWAYS_INLINE void testingEncodeInPlace(const std::vector<T>& data) {
-  for (auto i = 0; i < data.size(); i++) {
-    PrefixSortEncoder::encode(data[i], (char*)data.data() + i * sizeof(T));
+/// Logic is as same as int32_t
+template <>
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    uint64_t value,
+    char* dest) const {
+  auto& v = *reinterpret_cast<uint64_t*>(dest);
+  v = __builtin_bswap64(value);
+  if (!ascending_) {
+    v = ~v;
   }
 }
 
-template <typename T>
-FOLLY_ALWAYS_INLINE void testingDecodeInPlace(const std::vector<T>& data) {
-  for (auto i = 0; i < data.size(); i++) {
-    PrefixSortEncoder::decode<T>((char*)data.data() + i * sizeof(T));
+template <>
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    int64_t value,
+    char* dest) const {
+  encodeNoNulls((uint64_t)(value ^ 1ull << 63), dest);
+}
+namespace detail {
+/// Convert double to a uint64_t, their value comparison semantics remain
+/// consistent.
+static FOLLY_ALWAYS_INLINE uint64_t encodeDouble(double value) {
+  // Zero is the smallest positive value.
+  if (value == 0) {
+    return 1ull << 63;
+  }
+  // Nan is max value.
```

Comment:
How do you make sure that this won't collide with some other value? i.e. what guarantees that non-NaN double value will never get converted to the same uint64_t as this?

See https://github.com/facebookincubator/velox/pull/8177

Reply:
![image](https://github.com/facebookincubator/velox/assets/13514695/b98cf861-f67d-449d-872f-33c8fa4fcf06)

1. noramal values` Exponent will never be FF
2. infinity`s fraction is zero : 7f80 0000 = infinity  ff80 0000 = −infinity

So I think FF FF FF FF and FF FF FF FF -1 do not conflict with any float number.

ref:
https://en.wikipedia.org/wiki/Single-precision_floating-point_format

Reply:
Got it.

> exponents of −127 (all 0s) and +128 (all 1s) are reserved for special numbers.

This will guarantee no collision, right? Would be nice to include this point in the doc. Thanks.